### ISSUE TITLE
Add ability to easily log stack traces

### ISSUE THREAD

**Issue by inconshreveable** [inconshreveable/log15#10]
Feedback from the mailing list:

> A "stack trace" can be useful sometimes. What about adding a special "stack" key to the log functions (or better, a logger.XxxStack functions) which store the callers in the context.'

It's a difficult API design problem:
- Adding new XxxStack functions is too onerous on the API. 
- A simple handler can't just add them because it would end up generating the stack from within the handler tree.
- Maybe a handler could add them but it would have to do something smart where it walks up the captured stack frames and removes each frame until it reaches the caller of a Logger interface method.

Most likely, the best option is that this should be a new top-level function that you would use as a context value that we would know how to format nicely for you:

```
logger.Error("something went horribly wrong", "stack", log.Stack())
```

Note that it wouldn't work well with log.Lazy though (you'd end up with frames from the handler tree in the traceback).

How is it formatted as output? Newline escaping will produce an unreadable mess when looking at log files or on the terminal. grohl does something interesting where it splits the
stack into multiple log messages which is interesting to consider:

```
t=12412323 lvl=error stack=First line of stack trace
t=12412323 lvl=error stack=Second line of stack trace
t=12412323 lvl=error stack=Third line of stack trace
```

Maybe this format could just be used for the TerminalFormat.

**Comment by msabramo** [comment]
I guess this never made its way from @ChrisHines’s fork to the canonical repo?

Too bad. It looks like there was a bunch of great work and collaboration here. 

Came here wanting to be able to print stack traces when an `Error` is logged. My naive approach of wrapping the log15 logger with my own method is no good because of the hard-coded caller depth of 2 in log15, resulting in it looking like all errors come from my wrapper function :smile:

Is there a recommended approach for doing this?

**Comment by msabramo** [comment]
Yikes. https://github.com/ChrisHines/log15 doesn't even exist anymore. I was going to see if maybe using that fork was an option. I guess not. 😄

**Comment by ChrisHines** [comment]
@msabramo This work was merged long ago in #21. It is available as [`CallerStackHandler`](https://godoc.org/github.com/inconshreveable/log15#CallerStackHandler) in the log15 package. A while later the stack capture and formatting code was extracted into a standalone repo at [github.com/go-stack/stack](https://github.com/go-stack/stack). After that package was well established we refactored log15 to use it and removed that part of the code from this repo in #79. But CallerStackHandler is still available.
